host/docs/uhd_usrp_probe.1
.TH "uhd_usrp_probe" 1 "3.7.0" UHD "User Commands"
.SH NAME
uhd_usrp_probe \- USRP Hardware Driver Peripheral Report Utility
.SH DESCRIPTION
Report detailed information on UHD-supported Software Radio Peripherals
attached by USB, network, or embedded configuration.
.LP
The UHD package is the universal hardware driver for Ettus Research products. The goal
is to provide a host driver and API for current and future Ettus Research products.
Users will be able to use the UHD driver standalone or with 3rd party applications.
.LP
Details include unit names, revision numbers, and available sensors on all attached
USRP motherboards and daughterboards.
.SH SYNOPSIS
.B  uhd_usrp_probe [OPTIONS]
.SH OPTIONS
.IP "Device Address Arguments:"
--args \fIarg\fR
.IP "Print a complete property tree:"
--tree
.IP "Query a string value from the properties tree:"
--string \fIarg\fR
.IP "Query an integer value from the properties tree:"
--int \fIarg\fR
.IP "Query a double value from the properties tree:"
--double \fIarg\fR
.IP "Spawn an interactive prompt to peek/poke registers on an RFNoC block:"
--interactive-reg-iface \fIarg\fR
.IP "Print the version string and exit:"
--version
.IP "This help information:"
--help
.SH IDENTIFYING USRP DEVICES
.sp
Devices are addressed through key/value string pairs.
These string pairs can be used to narrow down the search for a specific device or group of devices.
Most UHD utility applications and examples have a \-\-args parameter that takes a device address,
where the device address is expressed as a delimited string.
See the documentation in types/device_addr.hpp for reference.

Every device has several ways of identifying it on the host system.
.SS Identifying by hardware identifier
.sp

All USRP devices can be found through their hardware series identifier, which match to USRP
devices as follows:

Argument    |  Device

type=usrp1  |  USRP1

type=usrp2  |  USRP2, USRP N200, USRP N210

type=b100   |  USRP B100

type=b200   |  USRP B200, USRP B210

type=e3x0   |  USRP E310

type=x300   |  USRP X300, USRP X310


.SS Identifying by serial number

All USRP devices are given a unique serial number, which can be used to identify a device as follows:

serial=12345678

.SS Identifying by IP address

USRP2, USRP N200, USRP N210, USRP X300, and USRP X310 devices connected to the host machine can all be found through their
IP addresses, as follows:

addr=192.168.10.2

.SS Identifying by name

Users have the option of giving their USRP devices short names using the usrp_burn_mb_eeprom utility
in lib/uhd/utils. Devices that have been given a name can be identified as follows:

name=foo

.fi
.fi
.SH EXAMPLES
.SS Device discovery via command line
.sp
Device address arguments can be supplied to narrow the scope of the search.
.sp
.nf
.ft C
uhd_usrp_probe \-\-args="type=usrp1"

\-\- OR \-\-

uhd_usrp_probe \-\-args="serial=12345678"
.ft P
.fi
.IP "Find all devices available to this system:"
uhd_usrp_probe
.SH SEE ALSO
UHD documentation:
.B http://files.ettus.com/manual/
.LP
GR-UHD documentation:
.B http://gnuradio.org/doc/doxygen/page_uhd.html
.LP
Other UHD programs:
.sp
uhd_find_devices(1)
.SH AUTHOR
This manual page was written by Maitland Bottoms and Nicholas Corgan
for the Debian project (but may be used by others).
.SH COPYRIGHT
Copyright (c) 2010 Ettus Research LLC
.LP
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
.LP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.


CHANGELOG
Change Log for Releases
==============================

## 003.014.001.000
N320: Terminate the DAC when not transmitting
E320: Add support for rev E
E320: Added .gitignore for FPGA build products
X300: Add DPDK support
X300: add capability to flash NI-2974 FPGA
X300: Broke two critical timing paths in FPGA
X300: fixed udp WSA buffer size assignment issue
E310: Fix DRAM_TEST target build
B200: Add bootloader for FX3 (fix for B2xx failing to enumerate)
TwinRX: Expose charge pump current for LO2
TwinRX: Add low spur mode for LO2
TwinRX: increase rev c lo1 charge pump default value
TwinRX: Fix tick rate
Device3: Constraint send/recv_frame_size based on down/upstream MTU
Device3: Fix MTU and default frame sizes
RFNoC: Search all nodes for tick rates
RFNoC: Change default address for the reg readbacks
uhd_image_builder: Let the OOT module point to folders not named "rfnoc"
uhd_image_builder: Add --auto-inst-src
MPMD: Fix spurious reclaim call after unclaim
MPMD: Release resources on destruction
MPM: Add SW/HW compat
liberio: Release context holder on destruction of last liberio xport
transport: fixed a pre-mature buffer reset
nirio: Fix typo in nirio_zero_copy
GPSD: fix API for 'gps_read'
AD9361: Fix return values for tune and set_clock_rate
DUC/DDC: Fix phase reset and accumulation
cores: Use NSDMI consistently in ?x_dsp_core_3000.*
Logging: fix deadlock issue on Windows machines
Logging: Fix ANSI colour codes
Utils: Add X300/X310 reset utility
Docs: Improved Windows-related build instructions
Docs: Add link to README for building custom filesystems for N3xx
sim: Fixing the port number in use for connection
tools: Fix for proj creation in ip_utils for tcl-based IP
tests: Fix mock_ctrl_iface for 32-bit MSVC

## 003.014.000.000
* N320: Add support for N320 and N321
* USRP-2974: Add support for USRP-2974
* DPDK: Add DPDK-based sockets-like library (for N3xx)
* N3xx: clocking API changes for transitioning clock and time sources
* N3xx: Bump max rev to G/6
* N3xx: Improve error messages for invalid clock/time settings
* N3xx: Get RFNoC crossbar baseport from FPGA
* N3xx: init peripherals before loading FPGA (to fix SFP0 init issues)
* N3xx: Move Linux kernel to 4.15
* N3xx/E320: Prepend SDK filename with device name
* N3xx: Update max rev to 7 (H)
* N3xx: Remove DDR3 from standard BIST collection
* N3xx: BIST: Improve DDR3 BIST to check for DmaFIFO
* N3xx: BIST: Auto-load the AA image for the ddr3 BIST
* N3xx: BIST: DDR3 test only enumerates first block
* N310: Modify AD9371 reset function to keep it in reset
* N310: move init_rf_cal before JESD de/framer bringup
* N310: Fix sporadic power on failures (requires firmware update)
* E3xx: Increase spp limit for E3xx radio
* E320: bist: Fix ref_clock lock test implementation
* E320: bist: Add link_up test
* E320: Add all 5 temp sensors, fan sensor and rssi sensors per channel
* E320: Fix tx/rx atr - antenna and frequency settings
* E320: Enable devtest for E320
* E320: images: Separate images package for Aurora image
* E320: Get RFNoC crossbar baseport from FPGA
* E320: add fpga_version_hash to e320 device info
* E310: Fix initialization of antenna and frequency values
* E31x: Destruct RFNoC before loading idle image
* X300: Reduce default send_frame_size to 4000 over Ethernet
* X300: Change Ethernet buffering
* X300: Log git hash and compat number as debug message
* X300: Move defaults to their own header
* X300: Use constrained_args
* X300: Enable clock_source and time_source device args
* X300: NIRIO: Demote RPC client cancel/abort to TRACE
* X300: remove default_buff_args properties
* X300: Remove 120 MHz master_clock_rate option
* X300: Set minimum master clock rate to 184.32 MHz
* X300: Factor our PID -> MB type and MB type -> product name mapping
* X300: Remove usage of boost::bind
* X300: Fix compiler warnings related to type conversions
* X300: Fix tick and sample rate setting
* X300: Enable ADC gain through RFNoC API
* X300: Demote NIRIO rpc client start/stop log messages to DEBUG
* X300: Enable 11.52 MHz and 23.04 MHz system ref rates
* X300: Enable x300_device_args.to_string()
* X300: Catch more inconsistencies in x300_device_args
* X300: Removed invalid 200 MHz sysref rate
* X300: Change PLL CP currents in x300_clock_ctrl
* B200: Remove superfluous fake lambda
* B200: Add support for user regs
* B200: Fix compiler warnings related to type conversions
* B100: Move fifo_ctrl_excelsior to b100 subdir
* B100: Fix fifo_ctrl_excelsior not exiting
* B100: Remove all Boostisms from fifo_ctrl_excelsior
* B100: Demote some clocking-related log messages to trace
* MPM: Get list of temperatures from all thermal zones
* MPM: add link_speed xport_info
* MPM: Add __mpm_device__ as usrp_hwd module variable
* MPM: Add usrp_update_fs
* MPM: Add i2c APIs for simple transfers
* MPM: Add vector-based transfer function for i2c
* MPM: Add variable configuration support to nijesdcore
* MPM: Add eyescan utility to nijesdcore
* MPM: Add PRBS-31 testing to nijesdcore
* MPM: Add convenience function to pull i2c bus from device tree
* MPM: Open and close i2c file descriptor on every access
* MPM: Multiprocessing instead of threading for claimer loop
* MPM: Factor out user EEPROM code into own module
* MPM: Add gpgga sensor function to GPSd iface
* MPM: Add bridge mode support
* MPM: Parameterize max UDP link allocation
* MPM: xport: add commit_xport docstring
* MPM: Improve error message on double-claim
* MPMD: Parallelize broadcast-finding
* MPMD: add option to enum rfnoc blocks from args
* MPMD: add link speed to xport udp
* MPMD: Add API to set RPC timeout atomically
* MPMD: Move timeout constants to header
* MPMD: Use new RPC API with timeout
* MPMD: Increase claim_rpc call timeout
* MPMD: implement get_*x_hints

host/utils/uhd_usrp_probe.cpp
//
// Copyright 2010-2011,2015-2016 Ettus Research LLC
// Copyright 2018 Ettus Research, a National Instruments Company
//
// SPDX-License-Identifier: GPL-3.0-or-later
//

#include <uhd/device.hpp>
#include <uhd/device3.hpp>
#include <uhd/property_tree.hpp>
#include <uhd/types/ranges.hpp>
#include <uhd/types/sensors.hpp>
#include <uhd/usrp/dboard_eeprom.hpp>
#include <uhd/usrp/dboard_id.hpp>
#include <uhd/usrp/mboard_eeprom.hpp>
#include <uhd/utils/safe_main.hpp>
#include <uhd/utils/cast.hpp>
#include <uhd/rfnoc_graph.hpp>
#include <uhd/rfnoc/block_id.hpp>
#include <uhd/version.hpp>
#include <boost/algorithm/string.hpp> //for split
#include <boost/format.hpp>
#include <boost/program_options.hpp>
#include <boost/lexical_cast.hpp>
#include <cstdlib>
#include <iostream>
#include <sstream>
#include <vector>

namespace po = boost::program_options;
using namespace uhd;

static std::string make_border(const std::string& text)
{
    std::stringstream ss;
    ss << boost::format("  _____________________________________________________")
       << std::endl;
    ss << boost::format(" /") << std::endl;
    std::vector<std::string> lines;
    boost::split(lines, text, boost::is_any_of("\n"));
    while (lines.back().empty())
        lines.pop_back(); // strip trailing newlines
    if (lines.size())
        lines[0] = "    " + lines[0]; // indent the title line
    for (const std::string& line : lines) {
        ss << boost::format("|   %s") % line << std::endl;
    }
    // ss << boost::format(" \\_____________________________________________________") <<
    // std::endl;
    return ss.str();
}

static std::string get_dsp_pp_string(
    const std::string& type, property_tree::sptr tree, const fs_path& path)
{
    std::stringstream ss;
    ss << boost::format("%s DSP: %s") % type % path.leaf() << std::endl;
    ss << std::endl;
    meta_range_t freq_range = tree->access<meta_range_t>(path / "freq/range").get();
    ss << boost::format("Freq range: %.3f to %.3f MHz") % (freq_range.start() / 1e6)
              % (freq_range.stop() / 1e6)
       << std::endl;
    ;
    return ss.str();
}

static std::string prop_names_to_pp_string(const std::vector<std::string>& prop_names)
{
    std::stringstream ss;
    size_t count = 0;
    for (const std::string& prop_name : prop_names) {
        ss << ((count++) ? ", " : "") << prop_name;
    }
    return ss.str();
}

static std::string get_frontend_pp_string(
    const std::string& type, property_tree::sptr tree, const fs_path& path)
{
    std::stringstream ss;
    ss << boost::format("%s Frontend: %s") % type % path.leaf() << std::endl;
    // ss << std::endl;

    ss << boost::format("Name: %s") % (tree->access<std::string>(path / "name").get())
       << std::endl;
    ss << boost::format("Antennas: %s")
              % prop_names_to_pp_string(
                    tree->access<std::vector<std::string>>(path / "antenna/options")
                        .get())
       << std::endl;
    if (tree->exists(path / "sensors")) {
        ss << boost::format("Sensors: %s")
                  % prop_names_to_pp_string(tree->list(path / "sensors"))
           << std::endl;
    }

    meta_range_t freq_range = tree->access<meta_range_t>(path / "freq/range").get();
    ss << boost::format("Freq range: %.3f to %.3f MHz") % (freq_range.start() / 1e6)
              % (freq_range.stop() / 1e6)
       << std::endl;

    std::vector<std::string> gain_names = tree->list(path / "gains");
    if (gain_names.size() == 0)
        ss << "Gain Elements: None" << std::endl;
    for (const std::string& name : gain_names) {
        meta_range_t gain_range =
            tree->access<meta_range_t>(path / "gains" / name / "range").get();
        ss << boost::format("Gain range %s: %.1f to %.1f step %.1f dB") % name
                  % gain_range.start() % gain_range.stop() % gain_range.step()
           << std::endl;
    }
    if (tree->exists(path / "bandwidth" / "range")) {
        meta_range_t bw_range =
            tree->access<meta_range_t>(path / "bandwidth" / "range").get();
        ss << boost::format("Bandwidth range: %.1f to %.1f step %.1f Hz")
                  % bw_range.start() % bw_range.stop() % bw_range.step()
           << std::endl;
    }

    ss << boost::format("Connection Type: %s")
              % (tree->access<std::string>(path / "connection").get())
       << std::endl;
    ss << boost::format("Uses LO offset: %s")
              % ((tree->exists(path / "use_lo_offset")
                     and tree->access<bool>(path / "use_lo_offset").get())
                        ? "Yes"
                        : "No")
       << std::endl;

    return ss.str();
}

static std::string get_codec_pp_string(
    const std::string& type, property_tree::sptr tree, const fs_path& path)
{
    std::stringstream ss;
    if (tree->exists(path / "name")) {
        ss << boost::format("%s Codec: %s") % type % path.leaf() << std::endl;

        ss << boost::format("Name: %s") % (tree->access<std::string>(path / "name").get())
           << std::endl;
        std::vector<std::string> gain_names = tree->list(path / "gains");
        if (gain_names.size() == 0)
            ss << "Gain Elements: None" << std::endl;
        for (const std::string& name : gain_names) {
            meta_range_t gain_range =
                tree->access<meta_range_t>(path / "gains" / name / "range").get();
            ss << boost::format("Gain range %s: %.1f to %.1f step %.1f dB") % name
                      % gain_range.start() % gain_range.stop() % gain_range.step()
               << std::endl;
        }
    }
    return ss.str();
}

static std::string get_dboard_pp_string(
    const std::string& type, property_tree::sptr tree, const fs_path& path)
{
    std::stringstream ss;
    ss << boost::format("%s Dboard: %s") % type % path.leaf() << std::endl;
    // ss << std::endl;
    const std::string prefix = (type == "RX") ? "rx" : "tx";
    if (tree->exists(path / (prefix + "_eeprom"))) {
        usrp::dboard_eeprom_t db_eeprom =
            tree->access<usrp::dboard_eeprom_t>(path / (prefix + "_eeprom")).get();
        if (db_eeprom.id != usrp::dboard_id_t::none())
            ss << boost::format("ID: %s") % db_eeprom.id.to_pp_string() << std::endl;
        if (not db_eeprom.serial.empty())
            ss << boost::format("Serial: %s") % db_eeprom.serial << std::endl;
        if (type == "TX" and tree->exists(path / "gdb_eeprom")) {
            usrp::dboard_eeprom_t gdb_eeprom =
                tree->access<usrp::dboard_eeprom_t>(path / "gdb_eeprom").get();
            if (gdb_eeprom.id != usrp::dboard_id_t::none())
                ss << boost::format("ID: %s") % gdb_eeprom.id.to_pp_string() << std::endl;
            if (not gdb_eeprom.serial.empty())
                ss << boost::format("Serial: %s") % gdb_eeprom.serial << std::endl;
        }
    }
    if (tree->exists(path / (prefix + "_frontends"))) {
        for (const std::string& name : tree->list(path / (prefix + "_frontends"))) {
            ss << make_border(get_frontend_pp_string(
                type, tree, path / (prefix + "_frontends") / name));
        }
    }
    ss << make_border(get_codec_pp_string(type,
        tree,
        path.branch_path().branch_path() / (prefix + "_codecs") / path.leaf()));
    return ss.str();
}

static std::string get_rfnoc_blocks_pp_string(rfnoc::rfnoc_graph::sptr graph)
{
    std::stringstream ss;
    ss << "RFNoC blocks on this device:" << std::endl << std::endl;
    for (const std::string& name : graph->find_blocks("")) {
        ss << "* " << name << std::endl;
    }
    return ss.str();
}

static std::string get_rfnoc_connections_pp_string(rfnoc::rfnoc_graph::sptr graph)
{
    std::stringstream ss;
    ss << "Static connections on this device:" << std::endl << std::endl;
    for (const auto& edge : graph->enumerate_static_connections()) {
        ss << "* " << edge.to_string() << std::endl;
    }
    return ss.str();
}

static std::string get_rfnoc_pp_string(rfnoc::rfnoc_graph::sptr graph)
{
    std::stringstream ss;
    ss << make_border(get_rfnoc_blocks_pp_string(graph));
    ss << make_border(get_rfnoc_connections_pp_string(graph));
    return ss.str();
}

static std::string get_mboard_pp_string(property_tree::sptr tree, const fs_path& path)
{
    std::stringstream ss;
    ss << boost::format("Mboard: %s") % (tree->access<std::string>(path / "name").get())
       << std::endl;

    if (tree->exists(path / "eeprom")) {
        usrp::mboard_eeprom_t mb_eeprom =
            tree->access<usrp::mboard_eeprom_t>(path / "eeprom").get();
        for (const std::string& key : mb_eeprom.keys()) {
            if (not mb_eeprom[key].empty())
                ss << boost::format("%s: %s") % key % mb_eeprom[key] << std::endl;
        }
    } else {
        ss << "No mboard EEPROM found." << std::endl;
    }
    if (tree->exists(path / "fw_version")) {
        ss << "FW Version: " << tree->access<std::string>(path / "fw_version").get()
           << std::endl;
    }
    if (tree->exists(path / "mpm_version")) {
        ss << "MPM Version: " << tree->access<std::string>(path / "mpm_version").get()
           << std::endl;
    }
    if (tree->exists(path / "fpga_version")) {
        ss << "FPGA Version: " << tree->access<std::string>(path / "fpga_version").get()
           << std::endl;
    }
    if (tree->exists(path / "fpga_version_hash")) {
        ss << "FPGA git hash: "
           << tree->access<std::string>(path / "fpga_version_hash").get() << std::endl;
    }
    if (tree->exists(path / "xbar")) {
        ss << "RFNoC capable: Yes" << std::endl;
    }
    ss << std::endl;
    try {
        if (tree->exists(path / "time_source" / "options")) {
            const std::vector<std::string> time_sources =
                tree->access<std::vector<std::string>>(path / "time_source" / "options")
                    .get();
            ss << "Time sources:  " << prop_names_to_pp_string(time_sources) << std::endl;
        }
        if (tree->exists(path / "clock_source" / "options")) {
            const std::vector<std::string> clock_sources =
                tree->access<std::vector<std::string>>(path / "clock_source" / "options")
                    .get();
            ss << "Clock sources: " << prop_names_to_pp_string(clock_sources)
               << std::endl;
        }
        if (tree->exists(path / "sensors")) {
            ss << "Sensors: " << prop_names_to_pp_string(tree->list(path / "sensors"))
               << std::endl;
        }
        if (tree->exists(path / "rx_dsps")) {
            for (const std::string& name : tree->list(path / "rx_dsps")) {
                ss << make_border(get_dsp_pp_string("RX", tree, path / "rx_dsps" / name));
            }
        }
        if (tree->exists(path / "dboards")) {
            for (const std::string& name : tree->list(path / "dboards")) {
                ss << make_border(
                    get_dboard_pp_string("RX", tree, path / "dboards" / name));
            }
            if (tree->exists(path / "tx_dsps")) {
                for (const std::string& name : tree->list(path / "tx_dsps")) {
                    ss << make_border(
                        get_dsp_pp_string("TX", tree, path / "tx_dsps" / name));
                }
            }
            for (const std::string& name : tree->list(path / "dboards")) {
                ss << make_border(
                    get_dboard_pp_string("TX", tree, path / "dboards" / name));
            }
        }
    } catch (const uhd::lookup_error& ex) {
        std::cout << "Exited device probe on " << ex.what() << std::endl;
    }
    return ss.str();
}


static std::string get_device_pp_string(property_tree::sptr tree)
{
    std::stringstream ss;
    ss << boost::format("Device: %s") % (tree->access<std::string>("/name").get())
       << std::endl;
    // ss << std::endl;
    for (const std::string& name : tree->list("/mboards")) {
        ss << make_border(get_mboard_pp_string(tree, "/mboards/" + name));
    }
    return ss.str();
}

void print_tree(const uhd::fs_path& path, uhd::property_tree::sptr tree)
{
    std::cout << path << std::endl;
    for (const std::string& name : tree->list(path)) {
        print_tree(path / name, tree);
    }
}

namespace {

    uint32_t str2uint32(const std::string& str)
    {
        if (str.find("0x") == 0) {
            return cast::hexstr_cast<uint32_t>(str);
        }
        return boost::lexical_cast<uint32_t>(str);
    }

    void shell_print_help()
    {
        std::cout << "Commands:\n\n"
            << "poke32 $addr $data     : Write $data to $addr\n"
            << "peek32 $addr           : Read from $addr and print\n"
            << "help                   : Show this\n"
            << "quit                   : Terminate shell\n"
            << std::endl;
    }

    void run_interactive_regs_shell(rfnoc::noc_block_base::sptr blk_ctrl)
    {
        std::cout << "<<< Interactive Block Peeker/Poker >>>" << std::endl;
        std::cout << "Type 'help' to get a list of commands." << std::endl;
        while (true) {
            std::string input;
            std::cout << ">>> " <<std::flush;
            std::getline(std::cin, input);
            std::stringstream ss(input);
            std::string command;
            ss >> command;
            if (command == "poke32") {
                std::string addr_s, data_s;
                uint32_t addr, data;
                try {
                    ss >> addr_s >> data_s;
                    addr = str2uint32(addr_s);
                    data = str2uint32(data_s);
                } catch (std::exception&) {
                    std::cout << "Usage: poke32 $addr $data" << std::endl;
                    continue;
                }
                blk_ctrl->regs().poke32(addr, data);
            }
            if (command == "peek32") {
                std::string addr_s;
                uint32_t addr;
                try {
                    ss >> addr_s;
                    addr = str2uint32(addr_s);
                } catch (std::exception&) {
                    std::cout << "Usage: peek32 $addr" << std::endl;
                    continue;
                }
                std::cout << "==> " << std::hex << blk_ctrl->regs().peek32(addr)
                          << std::dec << std::endl;
            }

            if (input == "help") {
                shell_print_help();
            }
            if (input == "quit") {
                return;
            }
        }
    }

}

int UHD_SAFE_MAIN(int argc, char* argv[])
{
    po::options_description desc("Allowed options");
    // clang-format off
    desc.add_options()
        ("help", "help message")
        ("version", "print the version string and exit")
        ("args", po::value<std::string>()->default_value(""), "device address args")
        ("tree", "specify to print a complete property tree")
        ("string", po::value<std::string>(), "query a string value from the property tree")
        ("double", po::value<std::string>(), "query a double precision floating point value from the property tree")
        ("int", po::value<std::string>(), "query a integer value from the property tree")
        ("sensor", po::value<std::string>(), "query a sensor value from the property tree")
        ("range", po::value<std::string>(), "query a range (gain, bandwidth, frequency, ...)  from the property tree")
        ("vector", "when querying a string, interpret that as std::vector")
        ("init-only", "skip all queries, only initialize device")
        ("interactive-reg-iface", po::value<std::string>(), "RFNoC devices only: Spawn a shell to interactively peek and poke registers on RFNoC blocks")
    ;
    // clang-format on

    po::variables_map vm;
    po::store(po::parse_command_line(argc, argv, desc), vm);
    po::notify(vm);

    // print the help message
    if (vm.count("help")) {
        std::cout << boost::format("UHD USRP Probe %s") % desc << std::endl;
        return EXIT_FAILURE;
    }

    if (vm.count("version")) {
        std::cout << uhd::get_version_string() << std::endl;
        return EXIT_SUCCESS;
    }

    device::sptr dev         = device::make(vm["args"].as<std::string>());
    property_tree::sptr tree = dev->get_tree();
    rfnoc::rfnoc_graph::sptr graph;
    try {
        graph = rfnoc::rfnoc_graph::make(vm["args"].as<std::string>());
    } catch (uhd::key_error&) {
        // pass
    }

    if (vm.count("string")) {
        if (vm.count("vector")) {
            std::vector<std::string> str_vector =
                tree->access<std::vector<std::string>>(vm["string"].as<std::string>())
                    .get();
            std::cout << "(";
            for (const std::string& str : str_vector) {
                std::cout << str << ",";
            }
            std::cout << ")" << std::endl;
        } else {
            std::cout << tree->access<std::string>(vm["string"].as<std::string>()).get()
                      << std::endl;
        }
        return EXIT_SUCCESS;
    }

    if (vm.count("double")) {
        std::cout << tree->access<double>(vm["double"].as<std::string>()).get()
                  << std::endl;
        return EXIT_SUCCESS;
    }

    if (vm.count("int")) {
        std::cout << tree->access<int>(vm["int"].as<std::string>()).get() << std::endl;
        return EXIT_SUCCESS;
    }

    if (vm.count("sensor")) {
        std::cout << tree->access<uhd::sensor_value_t>(vm["sensor"].as<std::string>())
                         .get()
                         .value
                  << std::endl;
        return EXIT_SUCCESS;
    }

    if (vm.count("range")) {
        meta_range_t range =
            tree->access<meta_range_t>(vm["range"].as<std::string>()).get();
        std::cout << boost::format("%.1f:%.1f:%.1f") % range.start() % range.step()
                         % range.stop()
                  << std::endl;
        return EXIT_SUCCESS;
    }

    if (vm.count("interactive-reg-iface")) {
        if (!graph) {
            std::cout << "ERROR: --interactive-reg-iface requires an RFNoC device!"
                      << std::endl;
            return EXIT_FAILURE;
        }
        const rfnoc::block_id_t block_id(vm["interactive-reg-iface"].as<std::string>());
        auto block_ctrl = graph->get_block(block_id);
        run_interactive_regs_shell(block_ctrl);
        return EXIT_SUCCESS;
    }

    if (vm.count("tree") != 0) {
        print_tree("/", tree);
    } else if (not vm.count("init-only")) {
        std::string device_pp_string = get_device_pp_string(tree);
        if (graph) {
            device_pp_string += get_rfnoc_pp_string(graph);
        }
        std::cout << make_border(device_pp_string) << std::endl;
    }

    return EXIT_SUCCESS;
}
